Dockerfile

```
@@ -1,36 +1,76 @@
-      python3-dev \
-      libffi-dev \
-      openssl-dev \
-      musl-dev \
-      cargo; \
-    pip3 install --upgrade pip cffi; \
-    pip3 install ansible boto pywinrm; \
-    apk del build-dependencies; \
-    rm -rf /var/cache/apk/*; \
-    mkdir -p /etc/ansible; \
-    echo -e "[local]\nlocalhost ansible_connection=local" > /etc/ansible/hosts
+FROM alpine:latest as runtime
 
+# Packages required for running Ansible and this resource
+RUN apk add --update --no-cache \
+		openssl ca-certificates \
+		python3 py3-pip \
+		ruby ruby-json \
+		openssh-client \
+		git
+	
+FROM runtime as builder
+
+# Packages required only to build and install cryptography
+RUN apk add --update --no-cache \
+		build-base \
+		python3-dev \
+		libffi-dev \
+		openssl-dev \
+		cargo \
+# Python packages required to run Ansible, and other nice-to-haves
+	&& pip3 install --no-cache-dir \
+		boto \
+		pywinrm \
+		ansible \
+# Clean up Python cache files
+	&& find /usr/lib/ -name '__pycache__' -print0 | xargs -0 -n1 rm -rf \
+	&& find /usr/lib/ -name '*.pyc' -print0 | xargs -0 -n1 rm -rf \
+
+# Add hosts for convenience
+	&& mkdir -p /etc/ansible \
+	&& echo -e "[local]\nlocalhost ansible_connection=local" > /etc/ansible/hosts
+
+# Copy Ruby resource scripts
 COPY assets/ /opt/resource/
 
-FROM main as testing
+FROM runtime as production
 
-RUN set -eux; \
-    gem install rspec; \
-    wget -q -O - https://raw.githubusercontent.com/troykinsella/mockleton/master/install.sh | bash; \
-    cp /usr/local/bin/mockleton /usr/local/bin/ansible-galaxy; \
-    cp /usr/local/bin/mockleton /usr/local/bin/ansible-playbook; \
-    cp /usr/local/bin/mockleton /usr/bin/ssh-add;
+# Copy and link Ansible and its dependencies
+RUN ln -s ansible /usr/bin/ansible-config \
+	&& ln -s ansible /usr/bin/ansible-console \
+	&& ln -s ansible /usr/bin/ansible-doc \
+	&& ln -s ansible /usr/bin/ansible-galaxy \
+	&& ln -s ansible /usr/bin/ansible-inventory \
+	&& ln -s ansible /usr/bin/ansible-playbook \
+	&& ln -s ansible /usr/bin/ansible-pull \
+	&& ln -s ansible /usr/bin/ansible-vault
+COPY --from=builder /usr/lib/python3.8/site-packages/ /usr/lib/python3.8/site-packages/
+COPY --from=builder /usr/bin/ansible /usr/bin/ansible
+COPY --from=builder /usr/bin/ansible-connection /usr/bin/ansible-connection
+COPY --from=builder /usr/bin/ansible-test /usr/bin/ansible-test
+COPY --from=builder /etc/ansible/ /etc/ansible/
 
-COPY . /resource/
+# Copy Ruby resource scripts
+COPY --from=builder /opt/resource/ /opt/resource/
 
+FROM production as testing
+
+# Sanity: ensure we can run Ansible
+RUN ansible-playbook --version
+
+# Install mockleton
+RUN apk add --update --no-cache bash
+RUN gem install rspec
+RUN wget -q -O - https://raw.githubusercontent.com/troykinsella/mockleton/master/install.sh | bash
+# Mock binaries used in testing
+RUN cp /usr/local/bin/mockleton /usr/bin/ansible-galaxy
+RUN cp /usr/local/bin/mockleton /usr/bin/ansible-playbook
+RUN cp /usr/local/bin/mockleton /usr/bin/ssh-add
```

Would it not be better to consolidate these commands into a single `RUN`?  Or is it better to keep them separate with this being a testing stage?  I'm somewhat new to multi-stage builds and don't use them regularly so forgive me if I don't know what I'm talking about 😄 

```suggestion
RUN echo 'Sanity: ensure we can run Ansible' \
  && ansible-playbook --version \
  && echo 'Install mockleton' \
  && apk add --update --no-cache bash \
  && gem install rspec \
  && wget -q -O - https://raw.githubusercontent.com/troykinsella/mockleton/master/install.sh | bash \
  && echo 'Mock binaries used in testing' \
  && cp /usr/local/bin/mockleton /usr/bin/ansible-galaxy \
  && cp /usr/local/bin/mockleton /usr/bin/ansible-playbook \
  && cp /usr/local/bin/mockleton /usr/bin/ssh-add
```